### PR TITLE
Reloading remote schema

### PR DESCRIPTION
**Pull request by angelozerr**
Reloading remote schema

See redhat-developer/vscode-xml#284

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
@BoykoAlex I'm starting to use your custom XML command support. I would like to provide some abstract class that you could use. Please tell me if you could use it. I mean https://github.com/eclipse/lemminx/pull/911/files#diff-ce0efddeef41b8881cd864523bdbfa55f5a1e310967e6b1014fce2f1fa26e273 which is an abstract command handler class which get a DOM Document from the first arguments where uri is filled with TextDocumentIdentifier.

Any feedback are welcome and it should be very nice if you could use it. Thanks!

**Comment by angelozerr** [comment]
See PR https://github.com/redhat-developer/vscode-xml/pull/343 on vscode-xml side.

**Comment by BoykoAlex** [comment]
@angelozerr The command I'm registering at the moment is somewhat similar to `XMLValidationAllFilesCommand`... It's not per document... The command I'm registering lets the client to send classpath change message to the server (any project in the workspace classpath change event).
However, if the command had a doc id as a parameter I'd go with `AbstractDOMDocumentCommandHandler`

**Comment by angelozerr** [comment]
This PR provides 2 XML custom commands :

 * a command to validate a given XML file which evicts cache for referenced grammars and retrigger the validation
 * a command to validate all opened XML files which evict the cache.

@datho7561 you can play with this PR and redhat-developer/vscode-xml#343 it should work.

I need to write tests and improve this PR to manage dependencies of referenced grammar to evict cache for those dependencies.

**Comment by datho7561** [comment]
It seems to work during my manual testing. I think that it makes sense to do the work to evict include/imported grammars in a separate PR. Its a feature that probably won't be used as much.

**Comment by angelozerr** [comment]
>  Its a feature that probably won't be used as much.

I agree, user could use the other commands to evict the full cache.

**Comment by angelozerr** [comment]
@datho7561 the PR is not ready, tests has been written.

**Review by datho7561 — CHANGES_REQUESTED** [review]
There were a few spots where exceptions were thrown and it wasn't that clear what would cause them. Other than that, it looks good.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/ContentModelManager.java [hunk 1]
```diff
@@ -400,6 +402,62 @@ public void setUseCache(boolean useCache) {
 		}
 	}
 
+	/**
+	 * Remove the referenced grammar from the given document and clear the Xerces
+	 * grammar cache which stores the XSD, DTD grammar.
+	 *
+	 * @param document the DOM document
+	 *
+	 * @throws IOException
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/ContentModelManager.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/commands/ArgumentsUtils.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+	/**
+	 * Returns the object from the given index as a given class type and null
+	 * otherwise.
+	 * 
+	 * @param <T>    the class type
+	 * @param params the execute command parameters.
+	 * @param index  the index
+	 * @param clazz  the class type.
+	 * @return the object from the given index as a given class type and null
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/extensions/commands/ArgumentsUtils.java:50]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/uriresolver/CacheResourcesManager.java [hunk 3]
```diff
@@ -275,6 +279,20 @@ public boolean isUseCache() {
 		return useCache;
 	}
 
+	/**
+	 * Remove the cache directory (.lemminx/cache) if it exists.
+	 * 
+	 * @throws IOException
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/uriresolver/CacheResourcesManager.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/uriresolver/XMLCacheResolverExtension.java [hunk 4]
```diff
@@ -135,4 +135,12 @@ public boolean isUseCache() {
 		return cacheResourcesManager.isUseCache();
 	}
 
+	/**
+	 * Remove the cache directory (.lemminx/cache) if it exists.
+	 * 
+	 * @throws IOException
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/uriresolver/XMLCacheResolverExtension.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/ContentModelManager.java [hunk 5]
```diff
@@ -400,6 +402,62 @@ public void setUseCache(boolean useCache) {
...
+				grammarPool.clear();
+			}
+		}
+	}
+
+	/**
+	 * Remove the cache directory (.lemminx/cache) if it exists and clear the Xerces
+	 * grammar cache which stores the XSD, DTD grammars.
+	 *
+	 * @throws IOException
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/ContentModelManager.java]: fixed

**Review by datho7561 — APPROVED** [review]

**Pull request merged**